src/core/core.datasetController.js

```diff
@@ -842,72 +853,147 @@ helpers.extend(DatasetController.prototype, {
-			dataset: me.getDataset(),
-			datasetIndex: me.index
-		};
-		const info = {cacheable: true};
-		let keys, i, ilen, key;
+		const context = me._getContext(index, active);
+		const info = {cacheable: !active};
+		let keys, i, ilen, key, value, readKey;
 
 		if (helpers.isArray(elementOptions)) {
 			for (i = 0, ilen = elementOptions.length; i < ilen; ++i) {
 				key = elementOptions[i];
-				values[key] = resolve([
-					datasetOpts[key],
-					options[key]
+				readKey = active ? 'hover' + key.charAt(0).toUpperCase() + key.slice(1) : key;
+				value = resolve([
+					datasetOpts[readKey],
+					options[readKey]
 				], context, index, info);
+				if (value !== undefined) {
+					values[key] = value;
+				}
 			}
 		} else {
 			keys = Object.keys(elementOptions);
 			for (i = 0, ilen = keys.length; i < ilen; ++i) {
 				key = keys[i];
-				values[key] = resolve([
-					datasetOpts[elementOptions[key]],
-					datasetOpts[key],
-					options[key]
+				readKey = active ? 'hover' + key.charAt(0).toUpperCase() + key.slice(1) : key;
+				value = resolve([
+					datasetOpts[elementOptions[readKey]],
+					datasetOpts[readKey],
+					options[readKey]
 				], context, index, info);
+				if (value !== undefined) {
+					values[key] = value;
+				}
 			}
 		}
 
 		if (info.cacheable) {
-			me._cachedDataOpts = Object.freeze(values);
+			values.$shared = true;
+			cached[mode] = values;
 		}
 
 		return values;
 	},
 
-	removeHoverStyle: function(element) {
-		helpers.merge(element._model, element.$previousStyle || {});
-		delete element.$previousStyle;
+	/**
+	 * @private
+	 */
+	_resolveAnimations: function(index, mode, active) {
+		const me = this;
+		const chart = me.chart;
+		const cached = me._cachedAnimations;
+		mode = mode || 'default';
+
+		if (cached[mode]) {
+			return cached[mode];
+		}
+
+		const info = {cacheable: true};
+		const context = me._getContext(index, active);
+		const datasetAnim = resolve([me._config.animation], context, index, info);
+		const chartAnim = resolve([chart.options.animation], context, index, info);
+		let config = helpers.mergeIf({}, [datasetAnim, chartAnim]);
+
+		if (active && config.active) {
+			config = helpers.extend({}, config, config.active);
+		}
+		if (mode === 'resize' && config.resize) {
+			config = helpers.extend({}, config, config.resize);
+		}
+
+		const animations = new Animations(chart, config);
+
+		if (info.cacheable) {
+			cached[mode] = animations && Object.freeze(animations);
+		}
+
+		return animations;
 	},
 
-	setHoverStyle: function(element, datasetIndex, index) {
-		const dataset = this.chart.data.datasets[datasetIndex];
-		const model = element._model;
-		const getHoverColor = helpers.getHoverColor;
+	/**
+	 * Utility for checking if the options are shared and should be animated separately.
+	 * @private
+	 */
+	_getSharedOptions: function(mode, element, options) {
+		if (mode !== 'reset' && options && options.$shared &&
+			element && element.options && element.options.$shared) {
```

Comment:
i'd indent this line one more tab

Reply:
Renamed `element` to `el` for (almost) fitting to one line instead.